docs/contributing/MODEL_MANAGER.md

```diff
@@ -0,0 +1,1214 @@
+mechanism.
+
+If successful, the method will return the appropriate subclass of
+`ModelConfigBase`. It will raise a `DuplicateModelException` if a
+model with the same key is already in the database, or an
+`InvalidModelConfigException` if a dict was passed and Pydantic
+experienced a parse or validation error.
+
+### update_model(key, config) -> AnyModelConfig:
+
+Given a key and a configuration, this will update the model
+configuration record in the database. `config` can be either a
+instance of `ModelConfigBase`, or a sparse `dict` containing the
+fields to be updated. This will return an `AnyModelConfig` on success,
+or raise `InvalidModelConfigException` or `UnknownModelException`
+exceptions on failure.
+
+***TO DO:*** Investigate why `update_model()` returns an
+`AnyModelConfig` while `add_model()` returns a `ModelConfigBase`.
+
+### rename_model(key, new_name) -> ModelConfigBase:
+
+This is a special case of `update_model()` for the use case of
+changing the model's name. It is broken out because there are cases in
+which the InvokeAI application wants to synchronize the model's name
+with its path in the `models` directory after changing the name, type
+or base. However, when using the ModelRecordService directly, the call
+is equivalent to:
+
+```
+store.rename_model(key, {'name': 'new_name'})
+```
+
+***TO DO:*** Investigate why `rename_model()` is returning a
+`ModelConfigBase` while `update_model()` returns a `AnyModelConfig`.
+
+***
+
+## Let's get loaded, the lowdown on ModelLoadService
+
+The `ModelLoadService` is responsible for loading a named model into
+memory so that it can be used for inference. Despite the fact that it
+does a lot under the covers, it is very straightforward to use.
+
+An application-wide model loader is created at API initialization time
+and stored in
+`ApiDependencies.invoker.services.model_loader`. However, you can
+create alternative instances if you wish.
+
+### Creating a ModelLoadService object
+
+The class is defined in
+`invokeai.app.services.model_loader_service`. It is initialized with
+an InvokeAIAppConfig object, from which it gets configuration
+information such as the user's desired GPU and precision, and with a
+previously-created `ModelRecordServiceBase` object, from which it
+loads the requested model's configuration information.
+
+Here is a typical initialization pattern:
+
+```
+from invokeai.app.services.config import InvokeAIAppConfig
+from invokeai.app.services.model_record_service import ModelRecordServiceBase
+from invokeai.app.services.model_loader_service import ModelLoadService
+
+config = InvokeAIAppConfig.get_config()
+store = ModelRecordServiceBase.open(config)
+loader = ModelLoadService(config, store)
+```
+
+Note that we are relying on the contents of the application
+configuration to choose the implementation of
+`ModelRecordServiceBase`.
+
+### get_model(key, [submodel_type], [context]) -> ModelInfo:
```

Comment:
This is a bit of a nit-pick, but I'd propose changing this API to:
```python
def get_model(self, key, context=None, **kwargs) -> ModelInfo:
```

`submodel_type` is only relevant for a small subset of models, so it seems wrong to include it in the service-level API. It can be passed forwarded as a kwarg, and then model-specific `get_model(...)` implementations can expect it as a required argument.